lib/build-components/build_components/__main__.py

```diff
@@ -0,0 +1,511 @@
+
+    The resulting image will be tagged as ``{project}-{component}:editor``
+
+    :param build_info: The build info populated from the config and command line arguments.
+    """
+    logger.info("Building editor image for %s component", build_info.component)
+
+    # Perform a "no context" docker build
+    _docker_build(
+        tags=build_info.editor_tag,
+        dockerfile=textwrap.dedent(
+            f"""
+            FROM {build_info.tag}
+            CMD "/bin/sh"
+            ENTRYPOINT []
+            """
+        ).encode(),
+    )
+
+
+def _docker_build(
+    tags: typing.Union[str, typing.List[str]],
+    buildargs: dict = None,
+    dockerfile: typing.Union[pathlib.Path, bytes] = None,
+) -> None:
+    """
+    A convenience wrapper for calling out to "docker build".
+
+    We always send the same context: the entire components/ directory.
+
+    :param tags: The tags to be applied to the built image.
+    :param buildargs: Values for ARG instructions in the dockerfile.
+    :param dockerfile: The dockerfile to build against. If not provided, it's assumed that a
+                       Dockerfile is present in the context directory. If it's a bytes object, it
+                       will be provided to the docker build process on stdin and a "no context"
+                       build will take place. Otherwise, a normal docker build will be performed
+                       with the specified Dockerfile.
+    """
+    buildargs = buildargs or {}
+    buildargs.setdefault("CACHE_BUST", str(time.time()))
+
+    cmd = ["env", "DOCKER_BUILDKIT=1", "docker", "build"]
+
+    for key, value in buildargs.items():
+        cmd.extend(["--build-arg", f"{key}={value}"])
+
+    tags = [tags] if isinstance(tags, str) else tags
+    for tag in tags:
+        cmd.extend(["--tag", tag])
+
+    if isinstance(dockerfile, bytes):
+        # If we receive the Dockerfile as content, we should pipe it to stdin.
+        # This is the "no context" build.
+        cmd.extend(["-"])
+    else:
+        # Otherwise, they can specify the path to the Dockerfile to use or let docker
+        # find one in the context directory.
+        if dockerfile:
+            cmd.extend(["-f", dockerfile.as_posix()])
+        cmd.extend(["components"])
+
+    logger.debug("Docker build command: %s", " ".join(cmd))
+    _check_call(cmd, stdin=dockerfile if isinstance(dockerfile, bytes) else None)
+
+
+def _check_call(cmd: typing.List[str], stdin: bytes = None) -> None:
+    """
+    Make a subprocess call and indent its output to match our python logging format.
+
+    :param cmd: The command to run.
+    :param stdin: Data to send to the subprocess as its input.
+    """
+    if stdin is None:
+        ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+        subprocess.run(["sed", "-e", "s/^/         /"], stdin=ps.stdout, check=True)
+        ps.wait()
+        if ps.returncode:
+            raise subprocess.CalledProcessError(ps.returncode, cmd)
+    else:
```

Comment:
it seems like there ought to be a way to combine most of the two branches of this conditional.  Also, it ought to be possible to hook the call to sed to the end of the pipeline before calling ps.communicate() in the second case (without having to read the output into memory then separately run it through sed).   Alternatively, you could possibly skip the pipeline altogether and just use textwrap.indent().

Reply:
It's been a while since I've been in this code, but I do remember trying a lot of things to minimize the duplicate code here. I wasn't very successful as you can tell. I was opting for `sed` due to its ability to stream the output as it occurred, which we'd lose with `indent()`. That being said, I think `communicate()` looses that anyway, so maybe `indent()` is preferable there if we can't otherwise combine these two paths a bit more.

Reply:
> without having to read the output into memory then separately run it through sed

Looking closer, I don't think that's what's happening. `stdout` is a file descriptor, not the contents being read completely into memory. The real sticking point is how to provide the input on `stdin`. `communicate()` was the best (only?) way I found to reasonably do that, And using that, forces us to call that prior to calling `run(['sed']) since the input there is the `stdout` from `communicate()`.

Reply:
...and I was mistaken. `communicate()` does in fact block and return the fully read-in stdout/stderr content. Still digging...

Reply:
i'm going to punt on this one. Python subprocess management doesn't seem to be designed to easily replicate bash shell behavior. I'd be glad to revisit this in a later PR, though.